client/src/controllers/UnsavedController.ts

```diff
@@ -181,6 +181,9 @@ export class UnsavedController extends Controller<HTMLFormElement> {
     if (!confirmationMessage) return null;
 
     if (this.forceValue || this.hasCommentsValue || this.hasEditsValue) {
+      const confirmEvent = this.dispatch('confirm', { cancelable: true });
+      if (confirmEvent.defaultPrevented) return null;
+
       // eslint-disable-next-line no-param-reassign
       event.returnValue = confirmationMessage;
```

Comment:
The way we trigger the browser confirmation dialog is actually deprecated. We should've done `event.preventDefault()` instead of setting `returnValue` on the event. See https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event

I wanted so badly to refactor this but I think it's out of scope for this PR... unless you want to review it too

Reply:
I've opened a new issue for this: #12132